plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -137,23 +147,42 @@ codeActionProvider ideState pId (CodeActionParams _ _ docId range _) = pluginRes
           title = "Expand record wildcard"
 
 collectRecordsRule :: Recorder (WithPriority Log) -> Rules ()
-collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp -> do
-  tmr <- use TypeCheck nfp
-  let exts = getEnabledExtensions <$> tmr
-      recs = concat $ maybeToList (getRecords <$> tmr)
-  logWith recorder Debug (LogCollectedRecords recs)
-  let renderedRecs = traverse renderRecordInfo recs
-      recMap = RangeMap.fromList (realSrcSpanToRange . renderedSrcSpan) <$> renderedRecs
-  logWith recorder Debug (LogRenderedRecords (concat renderedRecs))
-  pure ([], CRR <$> recMap <*> exts)
+collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp ->
+  justOrFail "Unable to TypeCheck" (use TypeCheck nfp) $ \tmr -> do
```

Comment:
Is using `fail` this way within a plugin rule acceptable? I didn't want to propagate the `Maybe` all the way to the end, especially when the plugin can't really do anything if it can't get a hold of the `TcModuleResult`. I am open to other suggestions regarding this.

Reply:
The `fail` used here makes use of `MonadFail Action`, which amounts to `MonadFail IO`. Since `fail @IO` just raises a user exception, this is a no-go then. I'll try to come up with an alternative solution.

Reply:
I'm not sure: we might catch it somewhere? I think it's probably bad. Maybe @pepeiborra knows? If it's not okay, we can potentially delete the `MonadFail Action` instance.

Reply:
I replaced the fail usage with more straightforward `Maybe` handling, but still it would be good to know if using the `MonadFail Action` instance is okay.

Reply:
Turns out that it's fine and exceptions will get turned into a diagnostic. Perhaps not a very good one, though. So up to you what you want to do here.